drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,668 @@
+                       ztimer_set(ZTIMER_USEC, &dev->ack_timer, LORA_ACK_REPLY_US);
+                    }
+                    else {
+                        DEBUG("[sx126x] RX frame doesn't require ACK frame.\n");
+                        hal->cb(hal, IEEE802154_RADIO_INDICATION_RX_DONE);
+                    }
+                }
+            else if (is_ack && !(dev->ack_filter)) {
+                    DEBUG("[sx126x] Received ACK.\n");
+                    hal->cb(hal, IEEE802154_RADIO_INDICATION_RX_DONE);
+                }
+                /* If all failed, simply drop the frame and continue listening
+                 * to incoming frames */
+                else {
+                    DEBUG("[sx126x] Addr filter failed or ACK filter on.\n");
+                    _set_state(dev, STATE_RX);
+                   
+                }        
+           
+    }
+    else if (irq_mask & SX126X_IRQ_PREAMBLE_DETECTED) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_PREAMBLE_DETECTED\n");
+    }
+    else if (irq_mask & SX126X_IRQ_SYNC_WORD_VALID) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_SYNC_WORD_VALID\n");
+        
+    }
+    else if (irq_mask & SX126X_IRQ_HEADER_VALID) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_HEADER_VALID\n");
+        hal->cb(hal, IEEE802154_RADIO_INDICATION_RX_START);
+    }
+    else if (irq_mask & SX126X_IRQ_HEADER_ERROR) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_HEADER_ERROR\n");
+    }
+    else if (irq_mask & SX126X_IRQ_CRC_ERROR) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_CRC_ERROR\n");
+        hal->cb(hal, IEEE802154_RADIO_INDICATION_CRC_ERROR);
+    }
+    else if (irq_mask & SX126X_IRQ_CAD_DONE) {
+        if (irq_mask & SX126X_IRQ_CAD_DETECTED){
+            DEBUG("[sx126x] netdev: SX126X_IRQ_CAD_DETECTED \n");
+            dev->cad_detected = true;
+        }
+        DEBUG("[sx126x] netdev: SX126X_IRQ_CAD_DONE\n");
+        hal->cb(hal, IEEE802154_RADIO_CONFIRM_CCA);
+        _set_state(dev, STATE_RX);
+        
+    }
+    else if (irq_mask & SX126X_IRQ_TIMEOUT) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_TIMEOUT\n");
+    }
+    else {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_NONE\n");
+    }
+
+    #if IS_USED(MODULE_SX126X_STM32WL)
+        NVIC_EnableIRQ(SUBGHZ_Radio_IRQn);
+    #endif  
+}
+}
+
+                        /*RADIO HAL FUNCTIONS*/
+
+static int _write(ieee802154_dev_t *hal, const iolist_t *iolist){
+
+    sx126x_t *dev = hal->priv;
+    (void)dev;
+    size_t pos = 0;
+    /* Full buffer used for Tx */
+    sx126x_set_buffer_base_address(dev, 0x80, 0x00);
+    /* Write payload buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        if (iol->iol_len > 0) {
+            sx126x_write_buffer(dev, pos + 0x80, iol->iol_base, iol->iol_len);
+            DEBUG("[sx126x] netdev: send: wrote data to payload buffer.\n");
+            pos += iol->iol_len;
+        }
+    }
+
+    sx126x_set_lora_payload_length(dev, pos);
```

Comment:
the current SubMAC assumes the radio adds the CRC.
This means, the `write` function should calculate it:
```diff
 static int _write(ieee802154_dev_t *hal, const iolist_t *iolist){
 
+
     sx126x_t *dev = hal->priv;
     (void)dev;
     size_t pos = 0;
+    uint16_t chksum = 0;
     /* Full buffer used for Tx */
     sx126x_set_buffer_base_address(dev, 0x80, 0x00);
     /* Write payload buffer */
     for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
         if (iol->iol_len > 0) {
             sx126x_write_buffer(dev, pos + 0x80, iol->iol_base, iol->iol_len);
+            memcpy(buffer + pos, iol->iol_base, iol->iol_len);
             DEBUG("[sx126x] netdev: send: wrote data to payload buffer.\n");
             pos += iol->iol_len;
+            chksum = ucrc16_calc_le(iol->iol_base, iol->iol_len,
+                                UCRC16_CCITT_POLY_LE, chksum);
         }
     }
-
+    chksum = byteorder_htols(chksum).u16;
+    /* Include CRC */
+    sx126x_write_buffer(dev, pos + 0x80, (uint8_t*) &chksum, sizeof(chksum));
+    memcpy(buffer + pos, &chksum, sizeof(chksum));
+    pos += 2;
     sx126x_set_lora_payload_length(dev, pos);
     DEBUG("[sx126x] writing (size: %d).\n", (pos));
     return 0;
 }
```